File: frontend/src/lib/api/agents.ts
// Dialog-based Agents API adapter.
// This adapter translates between the old Agent* types and the new Dialog* endpoints.
// Backend: GET/POST/PUT/DELETE /dialogs/

import { apiFetch } from "./client";
import { getMe } from "./auth";
import type {
  AgentCreate,
  AgentListResponse,
  AgentResponse,
  AgentUpdate,
  DialogCreate,
  DialogResponse,
  DialogUpdate,
  DialogListResponse,
} from "@/lib/types";

// TODO: Consider caching the user profile to avoid repeated fetches
let cachedUserId: string | null = null;

async function getUserId(): Promise<string | null> {
  if (cachedUserId) return cachedUserId;
  try {
    const user = await getMe();
    cachedUserId = user.user_id;
    return user.user_id;
  } catch {
    return null;
  }
}

// ── Adapter helpers ─────────────────────────────────────────────────────────────

function asString(value: unknown, fallback = ""): string {
  return typeof value === "string" ? value : fallback;
}

function asNumber(value: unknown, fallback: number): number {
  return typeof value === "number" ? value : fallback;
}

function dialogToAgent(dialog: DialogResponse): AgentResponse {
  // Extract embedding model from meta_data_filter if stored
  const meta = dialog.meta_data_filter || {};
  const promptConfig = dialog.prompt_config || {};
  const llmSetting = dialog.llm_setting || {};
  return {
    agent_id: dialog.id,
    name: dialog.name || "",
    description: dialog.description || "",
    config: {
      system_prompt: asString(promptConfig.system),
      dataset_id:
        Array.isArray(dialog.kb_ids) && dialog.kb_ids.length > 0
          ? dialog.kb_ids[0]
          : "",
      embedding_model: asString(meta.embedding_model),
      chat_model: dialog.llm_id,
      temperature: asNumber(llmSetting.temperature, 0.7),
      top_k: dialog.top_k ?? 5,
    },
    created_at: dialog.create_time
      ? new Date(dialog.create_time * 1000).toISOString()
      : new Date().toISOString(),
    updated_at: dialog.update_time
      ? new Date(dialog.update_time * 1000).toISOString()
      : new Date().toISOString(),
  };
}

function agentCreateToDialogCreate(payload: AgentCreate): DialogCreate {
  // Note: tenant_id will be set by the caller using the actual user ID
  // The DialogCreate type requires tenant_id, but we'll use a placeholder
  // that gets replaced before the API call
  return {
    tenant_id: "", // Will be filled by createAgent
    name: payload.name,
    description: payload.description || null,
    language: "English",
    llm_id: payload.config.chat_model,
    llm_setting: { temperature: payload.config.temperature },
    prompt_type: "simple",
    prompt_config: { system: payload.config.system_prompt },
    kb_ids: [payload.config.dataset_id],
    meta_data_filter: { embedding_model: payload.config.embedding_model },
    similarity_threshold: 0.2,
    vector_similarity_weight: 0.3,
    top_n: 6,
    top_k: payload.config.top_k,
    do_refer: "1",
    rerank_id: "",
    status: "1",
  };
}

function agentUpdateToDialogUpdate(
  agentId: string,
  payload: AgentUpdate,
): DialogUpdate {
  const update: DialogUpdate = {};
  if (payload.name !== undefined) update.name = payload.name;
  if (payload.description !== undefined)
    update.description = payload.description;
  if (payload.config) {
    const cfg = payload.config;
    if (cfg.chat_model !== undefined) update.llm_id = cfg.chat_model;
    if (cfg.temperature !== undefined)
      update.llm_setting = { temperature: cfg.temperature };
    if (cfg.system_prompt !== undefined)
      update.prompt_config = { system: cfg.system_prompt };
    if (cfg.dataset_id !== undefined) update.kb_ids = [cfg.dataset_id];
    if (cfg.embedding_model !== undefined)
      update.meta_data_filter = { embedding_model: cfg.embedding_model };
    if (cfg.top_k !== undefined) update.top_k = cfg.top_k;
  }
  return update;
}

// ── API functions ───────────────────────────────────────────────────────────────

/**
 * List all agents (dialogs).
 */
export async function listAgents(): Promise<AgentListResponse> {
  const res = await apiFetch<DialogListResponse>("/dialogs/");
  return {
    agents: res.dialogs.map(dialogToAgent),
    total: res.total,
  };
}

/**
 * Create a new agent (dialog).
 */
export async function createAgent(
  payload: AgentCreate,
): Promise<AgentResponse> {
  const tenantId = await getUserId();
  if (!tenantId) {
    throw new Error("Not authenticated. Please log in again.");
  }
  const dialogCreate = agentCreateToDialogCreate(payload);
  // Inject the actual tenant ID from the logged-in user
  dialogCreate.tenant_id = tenantId;
  const dialog = await apiFetch<DialogResponse>("/dialogs/", {
    method: "POST",
    body: JSON.stringify(dialogCreate),
  });
  return dialogToAgent(dialog);
}

/**
 * Get a single agent by ID.
 */
export async function getAgent(agentId: string): Promise<AgentResponse> {
  const dialog = await apiFetch<DialogResponse>(`/dialogs/${agentId}`);
  return dialogToAgent(dialog);
}

/**
 * Update an agent (dialog).
 */
export async function updateAgent(
  agentId: string,
  payload: AgentUpdate,
): Promise<AgentResponse> {
  const dialogUpdate = agentUpdateToDialogUpdate(agentId, payload);
  const dialog = await apiFetch<DialogResponse>(`/dialogs/${agentId}`, {
    method: "PUT",
    body: JSON.stringify(dialogUpdate),
  });
  return dialogToAgent(dialog);
}

/**
 * Delete an agent (dialog).
 */
export async function deleteAgent(agentId: string): Promise<void> {
  await apiFetch<void>(`/dialogs/${agentId}`, { method: "DELETE" });
}

File: backend/api/routes/agents.py
from fastapi import APIRouter, Depends, HTTPException, Request, status

from backend.api.dependencies import get_current_user
from backend.api.routes._limiter import limiter
from backend.models_peewee import User
from backend.schemas.agents import (
    AgentCreate,
    AgentListResponse,
    AgentResponse,
    AgentUpdate,
)
from backend.services.ingestion_factory import get_agent_store

router = APIRouter(prefix="/agents", tags=["agents"])


def _ensure_agent_access(agent, user: User) -> None:
    owner_id = getattr(agent, "owner_id", None)
    if user.is_admin or owner_id == user.id:
        return
    raise HTTPException(status_code=403, detail="You do not have access to this agent")


def _agent_to_response(agent) -> AgentResponse:
    from backend.schemas.agents import AgentConfigSchema

    return AgentResponse(
        agent_id=agent.agent_id,
        name=agent.name,
        description=agent.description,
        config=AgentConfigSchema(
            system_prompt=agent.config.system_prompt,
            dataset_id=agent.config.dataset_id,
            embedding_model=agent.config.embedding_model,
            chat_model=agent.config.chat_model,
            temperature=agent.config.temperature,
            top_k=agent.config.top_k,
        ),
        created_at=agent.created_at,
        updated_at=agent.updated_at,
    )


@router.post("/", response_model=AgentResponse, status_code=status.HTTP_201_CREATED)
@limiter.limit("60/minute")
async def create_agent(
    request: Request,
    payload: AgentCreate,
    user: User = Depends(get_current_user),
) -> AgentResponse:
    from rag.storage.agent_store import AgentConfig

    store = get_agent_store()
    config = AgentConfig(
        system_prompt=payload.config.system_prompt,
        dataset_id=payload.config.dataset_id,
        embedding_model=payload.config.embedding_model,
        chat_model=payload.config.chat_model,
        temperature=payload.config.temperature,
        top_k=payload.config.top_k,
    )

    agent = store.create(
        name=payload.name,
        description=payload.description,
        config=config,
        owner_id=user.id,
    )

    return _agent_to_response(agent)


@router.get("/", response_model=AgentListResponse)
@limiter.limit("120/minute")
async def list_agents(
    request: Request,
    user: User = Depends(get_current_user),
) -> AgentListResponse:
    store = get_agent_store()
    agents = store.list(owner_id=None if user.is_admin else user.id)
    return AgentListResponse(
        agents=[_agent_to_response(a) for a in agents],
        total=len(agents),
    )


@router.get("/{agent_id}", response_model=AgentResponse)
@limiter.limit("120/minute")
async def get_agent(
    request: Request,
    agent_id: str,
    user: User = Depends(get_current_user),
) -> AgentResponse:
    store = get_agent_store()
    agent = store.get(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail=f"Agent '{agent_id}' not found")
    _ensure_agent_access(agent, user)
    return _agent_to_response(agent)


@router.put("/{agent_id}", response_model=AgentResponse)
@limiter.limit("60/minute")
async def update_agent(
    request: Request,
    agent_id: str,
    payload: AgentUpdate,
    user: User = Depends(get_current_user),
) -> AgentResponse:
    store = get_agent_store()
    existing = store.get(agent_id)
    if not existing:
        raise HTTPException(status_code=404, detail=f"Agent '{agent_id}' not found")
    _ensure_agent_access(existing, user)

    updates: dict = {}
    if payload.name is not None:
        updates["name"] = payload.name
    if payload.description is not None:
        updates["description"] = payload.description
    if payload.config is not None:
        updates["config"] = {
            "system_prompt": payload.config.system_prompt,
            "dataset_id": payload.config.dataset_id,
            "embedding_model": payload.config.embedding_model,
            "chat_model": payload.config.chat_model,
            "temperature": payload.config.temperature,
            "top_k": payload.config.top_k,
        }

    agent = store.update(agent_id, updates)
    if not agent:
        raise HTTPException(status_code=404, detail=f"Agent '{agent_id}' not found")
    return _agent_to_response(agent)


@router.delete("/{agent_id}", status_code=status.HTTP_204_NO_CONTENT)
@limiter.limit("60/minute")
async def delete_agent(
    request: Request,
    agent_id: str,
    user: User = Depends(get_current_user),
) -> None:
    store = get_agent_store()
    existing = store.get(agent_id)
    if not existing:
        raise HTTPException(status_code=404, detail=f"Agent '{agent_id}' not found")
    _ensure_agent_access(existing, user)

    deleted = store.delete(agent_id)
    if not deleted:
        raise HTTPException(status_code=404, detail=f"Agent '{agent_id}' not found")
